### PR TITLE
Unit test for Magento\Fedex\Plugin\Block\Tracking\PopupDeliveryDate

### PR DESCRIPTION
**Pull request by karyna-t**
### Description (*)
Covers Magento\Fedex\Plugin\Block\Tracking\PopupDeliveryDate class with a unit test.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    public function testAfterFormatDeliveryDateTimeWithFedexCarrier()
+    {
+        /** @var Status|MockObject $trackingStatusMock */
+        $trackingStatusMock = $this->getMockBuilder(Status::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getCarrier'])
+            ->getMock();
+        $trackingStatusMock->expects($this::once())
+            ->method('getCarrier')
+            ->willReturn(\Magento\Fedex\Model\Carrier::CODE);
```
Comment (edenduong) [inline app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: Could you please `use Magento\Fedex\Model\Carrier;`  and change this line to `->willReturn(Carrier::CODE);` ?

**Review by edenduong — CHANGES_REQUESTED** [review]
Thank you for your contribution. Please check my comment.

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php (hunk 1 above)
Comment (karyna-t) [reply app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: Thank you for the notice, done.

Comment (edenduong) [reply app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: Thank you !

**Review by edenduong — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @edenduong, thank you for the review.
[ENGCOM-6669](https://jira.corp.magento.com/browse/ENGCOM-6669) has been created to process this Pull Request
:eight_spoked_asterisk: @edenduong, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            ->disableOriginalConstructor()
+            ->setMethods(['formatDeliveryDate', 'getTrackingInfo'])
+            ->getMock();
+        $subjectMock->expects($this->once())
+            ->method('getTrackingInfo')
+            ->willReturn([[$trackingStatusMock]]);
+        $subjectMock->expects($this->once())
+            ->method('formatDeliveryDate');
+
+        $this->plugin->afterFormatDeliveryDateTime($subjectMock, 'Test Result', '2020-02-02', '12:00');
```
Comment (lbajsarowicz) [inline app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: Extract values to `const STUB_DELIVERY_DATE` and `const STUB_DELIVERY_TIME` to show that actual value does not really matter.

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    public function testAfterFormatDeliveryDateTimeWithOtherCarrier()
+    {
+        /** @var Status|MockObject $trackingStatusMock */
+        $trackingStatusMock = $this->getMockBuilder(Status::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getCarrier'])
+            ->getMock();
+        $trackingStatusMock->expects($this::once())
+            ->method('getCarrier')
+            ->willReturn('not-fedex');
```
Comment (lbajsarowicz) [inline app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: Extract to `const STUB_CARRIER_CODE_NOT_FEDEX` to explain what is actual role of the value.

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * Test the method with a different carrier
+     */
+    public function testAfterFormatDeliveryDateTimeWithOtherCarrier()
+    {
+        /** @var Status|MockObject $trackingStatusMock */
+        $trackingStatusMock = $this->getMockBuilder(Status::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getCarrier'])
+            ->getMock();
```
Comment (lbajsarowicz) [inline app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: If this part of code is exactly the same for both of tests - just extract that to `setUp` or even better - create private method `getStatusMock` :-)

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        $trackingStatusMock->expects($this::once())
+            ->method('getCarrier')
+            ->willReturn('not-fedex');
+
+        /** @var Popup|MockObject $subjectMock */
+        $subjectMock = $this->getMockBuilder(Popup::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['formatDeliveryDate', 'getTrackingInfo'])
+            ->getMock();
```
Comment (lbajsarowicz) [inline app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: If this part of code is exactly the same for both of tests - just extract that to setUp or even better - create private method getPopupMock :-)

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        $subjectMock->expects($this->once())
+            ->method('getTrackingInfo')
+            ->willReturn([[$trackingStatusMock]]);
```
Comment (lbajsarowicz) [inline app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: If you move above mentioned to setUp, you can  put these 3 lines there, too.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Cosmetic changes improving readability and reusability of tests.

File: app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php (hunk 2 above)
Comment (karyna-t) [reply app/code/Magento/Fedex/Test/Unit/Plugin/Block/Tracking/PopupDeliveryDateTest.php]: Fixed, thanks

**Comment by karyna-t** [comment]
@lbajsarowicz thanks for your review! I've updated the PR.

**Comment by lbajsarowicz** [comment]
Great job! Thanks, and approval is already there after last change.

**Comment by karyna-tsymbal-atwix** [comment]
Lukazs could you please approve this PR because there is an approve from Eden only. Thanks

**Review by lbajsarowicz — APPROVED** [review]
Actually we should wait to Functional Tests be fixed.
But conditionally - based on experience I can say that Functional tests failure is not caused by the PR change, so I approve PR.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6669](https://jira.corp.magento.com/browse/ENGCOM-6669) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**